contracts/poap/src/contract.rs

```diff
@@ -0,0 +1,443 @@
+
+    // Validate the creator address
+    let creator = deps.api.addr_validate(&msg.event_info.creator)?;
+
+    // Check that start time is before the end time
+    if !msg.event_info.end_time.gt(&msg.event_info.start_time) {
+        return Err(ContractError::StartTimeAfterEndTime {
+            start: msg.event_info.start_time,
+            end: msg.event_info.end_time,
+        });
+    }
+
+    // Check that the start time is in the future
+    if !msg.event_info.start_time.gt(&env.block.time) {
+        return Err(ContractError::StartTimeBeforeCurrentTime {
+            current_time: env.block.time,
+            start_time: msg.event_info.start_time,
+        });
+    }
+
+    // Check that the end time is in the future
+    if !msg.event_info.end_time.gt(&env.block.time) {
+        return Err(ContractError::EndTimeBeforeCurrentTime {
+            current_time: env.block.time,
+            end_time: msg.event_info.end_time,
+        });
+    }
+
+    // Check that the poap uri is a valid IPFS url
+    let poap_uri = Url::parse(&msg.event_info.base_poap_uri)
+        .map_err(|_err| ContractError::InvalidPoapUri {})?;
+    if poap_uri.scheme() != "ipfs" {
+        return Err(ContractError::InvalidPoapUri {});
+    }
+
+    // Check that the event uri is a valid IPFS url
+    let event_uri =
+        Url::parse(&msg.event_info.event_uri).map_err(|_err| ContractError::InvalidEventUri {})?;
+    if event_uri.scheme() != "ipfs" {
+        return Err(ContractError::InvalidEventUri {});
+    }
+
+    // Check per address limit
+    if msg.event_info.per_address_limit == 0 {
+        return Err(ContractError::InvalidPerAddressLimit {});
+    }
+
+    let config = Config {
+        admin: admin.clone(),
+        minter: minter.clone(),
+        per_address_limit: msg.event_info.per_address_limit,
+        cw721_code_id: msg.cw721_code_id.u64(),
+        mint_enabled: false,
+    };
+    // Save the received event info.
+    CONFIG.save(deps.storage, &config)?;
+
+    let event_info = EventInfo {
+        creator: creator.clone(),
+        start_time: msg.event_info.start_time,
+        end_time: msg.event_info.end_time,
+        base_poap_uri: msg.event_info.base_poap_uri.clone(),
+        event_uri: msg.event_info.event_uri.clone(),
+    };
+    // Save the event info
+    EVENT_INFO.save(deps.storage, &event_info)?;
+
+    // Submessage to instantiate cw721 contract
+    let cw721_submessage = SubMsg::reply_on_success(
+        wasm_instantiate(
+            msg.cw721_code_id.into(),
+            &Cw721InstantiateMsg {
+                name: msg.cw721_initiate_msg.name,
+                symbol: msg.cw721_initiate_msg.symbol,
+                minter: env.contract.address.to_string(),
+            },
+            info.funds,
+            "poap cw721".to_string(),
+        )?,
+        INSTANTIATE_CW721_REPLY_ID,
+    );
+
+    Ok(Response::new()
+        .add_attribute("method", "instantiate")
```

Comment:
It would be better to build const value to replace attribute string, say, `"action"` into `ATTRIBUTE_ACTION`, `"creator"` into `ATTRIBUTE_CREATOR`.